Added support for cs3 api

dynamic-entropy opened the pull request:
CS3APIs are a set of interfaces used to communicate between various entities within science mesh. 

This PR aims to add compatibility within the gfal multiplexer to perform third party copies with science mesh endpoints running the interoperability platform reva. 

We extend the `http-plugin` with the ability to fetch and initialise `Davix::RequestParams` with `Davix::RequestProtocol::CS3` for transfer endpoints with `cs3:` prefix.
Relevant interfaces were added to `libdavix` in PR https://github.com/cern-fts/davix/pull/72

dynamic-entropy committed:
Added support for cs3 api

glpatcern commented on doc/examples/gfal_copy.c:
```diff
@@ -47,21 +47,24 @@ static void my_monitor_callback(gfalt_transfer_status_t h, const char* src,
 
 int main(int argc, char** argv)
 {
+    printf("Main");
```
I'd drop any unneeded code, like this...

mpatrascoiu replied:
I'm of the same opinion. This example file should be left as it was originally, except for the addition of the `\n` symbol in the print statement.

glpatcern commented on doc/examples/gfal_copy.c:
```diff
@@ -76,10 +79,9 @@ int main(int argc, char** argv)
     gfalt_set_timeout(params, 60, &error);                  // 60 seconds timeout
     gfalt_set_dst_spacetoken(params, "TOKEN", &error);      // Destination space token, support depends on the plugin (SRM, XROOTD do support this)
-    gfalt_set_replace_existing_file(params, FALSE, &error); // Just in case, do not overwrite
+    gfalt_set_replace_existing_file(params, TRUE, &error); // Just in case, do not overwrite
```
also this should be left as in the original code

glpatcern commented on doc/examples/gfal_copy.c:
```diff
@@ -91,5 +93,6 @@ int main(int argc, char** argv)
     // Release memory
     gfalt_params_handle_delete(params, &error);
     gfal2_context_free(context);
+    printf("Transfer Compelete :) \n");
```
...and this can be removed as well

glpatcern commented on doc/examples/gfal_copy.c:
```diff
@@ -47,21 +47,24 @@ static void my_monitor_callback(gfalt_transfer_status_t h, const char* src,
 
+    setenv("SRC_URL", source, 1);
+    setenv("DST_URL", destination, 1);
+
```
Here I would add a comment stating that
```
The authentication with Reva is yet to be properly designed. Nevertheless, there is a need to associate tokens to URLs given that a token issued for a destination must have write rights, whereas right now GFAL issues a `stat()` == `HEAD` request also to the destination without requiring write rights. To be further developed.
```

mpatrascoiu replied:
I would revert even this. It's too specific to a protocol of a particular plugin to be part of the main `gfal_copy` example.

glpatcern replied:
I see your point. But then we need to document these env vars are required by the plugin -> in Davix?

mpatrascoiu replied:
Either Davix or the [gfal_http_plugin.cpp](https://github.com/cern-fts/gfal2/blob/master/src/plugins/http/gfal_http_plugin.cpp) file.

glpatcern commented on .gitignore:
```diff
@@ -17,3 +17,6 @@ DartConfiguration.tcl
 
+#buid folder
+build
+
```
Not sure this is in line with how GFAL is compiled, up to @mpatrascoiu to approve it (apart the small typo).

mpatrascoiu replied:
Hello @glpatcern,

We can leave the build folder in there. Correct the typo, remove the extra trailing whiteline :slightly_smiling_face:

glpatcern reviewed (changes requested):
I suggested some changes also to ease reviewing/merging for @mpatrascoiu

dynamic-entropy commented:
Thankyou @glpatcern 😊 , I will make the required changes.

mpatrascoiu commented on src/plugins/http/gfal_http_copy.cpp:
```diff
@@ -136,7 +136,7 @@ static void set_copy_mode_from_urls(gfal2_context_t context, const char * src_ur
 static bool is_http_scheme(const char* url)
 {
-    const char *schemes[] = {"http:", "https:", "dav:", "davs:", "s3:", "s3s:", "gcloud:", "gclouds:", "swift:", "swifts:", NULL};
+    const char *schemes[] = {"http:", "https:", "dav:", "davs:", "s3:", "s3s:", "gcloud:", "gclouds:", "swift:", "swifts:", "cs3:", NULL};
```
If you're adding `cs3s` in [Davix](https://github.com/cern-fts/davix/pull/72#discussion_r700010963), make sure to add it also here.

mpatrascoiu reviewed (commented):
Looks good. Almost ready to be merged!

I've also spotted here the odd extra-whitespace here-and there.
Again, I believe it has to do with the editor used.

Cheers,  
Mihai

dynamic-entropy committed:
Made the following changes

.gitignore
- removed trailing newline
- corrected typo, build to build

doc/example/gfal_copy.c
- Dropped additional print statements
- Removed cs3 specific env var assignments

src/plugins/http/gfal_http_copy.cpp
- Added cs3 in schemes list

src/plugin/http/gfal_http_plugin.cpp
- Added comments noting need to manually set source and destination as env variables for the time being
- Added cs3 to the http_url checks

dynamic-entropy commented:
I have made the requested amendments. Thank you for going through the PR. 😊 
Please let me know if you want me to squash them into a single commit?

mpatrascoiu commented:
Many thanks! Looking great.

I've merged the changes into [Gfal2/develop](https://gitlab.cern.ch/dmc/gfal2/-/commit/9a671772).

Gfal2 project is hosted mainly on [CERN Gitlab](https://gitlab.cern.ch/dmc/gfal2), with Gothub acting as a mirror.
It's a shame there's no good way to have the merge request here look green and bring the changes to Gitlab.

Cheers,
Mihai
